packages/reason-css-parser/ppx/EmitType.re

```diff
@@ -0,0 +1,269 @@
+open Longident;
+exception Unsupported;
+module Make = (Ast_builder: Ppxlib.Ast_builder.S) => {
+  open Ast_builder;
+  open Css_spec_parser;
+
+  let create_variant_name = (type_name, name) =>
+    type_name ++ "__make__" ++ name;
+
+  let txt = txt => {Location.loc: Ast_builder.loc, txt};
+
+  let ldot = (module_name, name) =>
+    Ldot(Lident(module_name), name) |> txt |> pexp_ident;
+
+  let module_op = ldot("Op");
+  let module_values = ldot("Values");
+
+  let construct = (~expr=None, name) =>
+    pexp_construct(txt(Lident(name)), expr);
+
+  let kebab_case_to_snake_case = name =>
+    name |> String.split_on_char('-') |> String.concat("_");
+  let first_uppercase = name =>
+    (String.sub(name, 0, 1) |> String.uppercase_ascii)
+    ++ String.sub(name, 1, String.length(name) - 1);
+
+  let is_function = str => {
+    open String;
+    let length = length(str);
+    length >= 2 && sub(str, length - 2, 2) == "()";
+  };
+  let function_value_name = function_name => "function-" ++ function_name;
+  let property_value_name = property_name =>
+    is_function(property_name)
+      ? function_value_name(property_name) : "property-" ++ property_name;
+
+  let value_of_delimiter =
+    fun
+    | "+" => "cross"
+    | "-" => "dash"
+    | "*" => "asterisk"
+    | "/" => "bar"
+    | "@" => "at"
+    | "," => "comma"
+    | ";" => ""
+    | ":" => "doubledot"
+    | "." => "dot"
+    | "(" => "openparen"
+    | ")" => "closeparen"
+    | "[" => "openbracket"
+    | "]" => "closebracket"
+    | "{" => "opencurly"
+    | "}" => "closecurly"
+    | "^" => "caret"
+    | "<" => "lessthan"
+    | "=" => "equal"
+    | ">" => "biggerthan"
+    | "|" => "vbar"
+    | "~" => "tilde"
+    | "$" => "dollar"
+    | _ => "unknown";
+
+  let value_name_of_css = str =>
+    String.(
+      {
+        let length = length(str);
+        let str =
+          if (is_function(str)) {
+            let str = sub(str, 0, length - 2);
+            function_value_name(str);
+          } else {
+            str;
+          };
+        kebab_case_to_snake_case(str);
+      }
+    );
+
+  let abstract_type = name => {
+    let type_ =
+      type_declaration(
+        ~name=txt(name),
+        ~params=[],
+        ~cstrs=[],
+        ~private_=Private,
+        ~manifest=None,
+        ~kind=Ptype_abstract,
+      );
+    pstr_type(Recursive, [type_]);
+  };
+  // TODO: multiplier name
+  let rec variant_name = value => {
+    let value_name =
+      switch (value) {
+      | Terminal(Delim(name), _) => value_of_delimiter(name)
```

Comment:
Would be perfect to annotate the parsing equivalent of Terminal, Group and Combinator.

Reply:
Can you give an example? I could not see clearly what you mean